clients/drcachesim/scheduler/scheduler.cpp

```diff
@@ -1255,6 +1217,205 @@ scheduler_tmpl_t<RecordType, ReaderType>::read_traced_schedule()
+    std::vector<std::map<uint64_t, uint64_t>> time_tree(inputs_.size());
+    for (int input_idx = 0; input_idx < static_cast<input_ordinal_t>(inputs_.size());
+         ++input_idx) {
+        for (int sched_idx = 0;
+             sched_idx < static_cast<int>(input_sched[input_idx].size()); ++sched_idx) {
+            schedule_input_tracker_t &sched = input_sched[input_idx][sched_idx];
+            VPRINT(this, 4, "as-read: input=%d start=%" PRId64 " time=%" PRId64 "\n",
+                   input_idx, sched.start_instruction, sched.timestamp);
+            time_tree[input_idx][sched.timestamp] = sched.start_instruction;
+        }
+    }
+
+    // Finally, convert the requested time ranges into instr ordinal ranges.
+    for (const auto &tid_it : workload_tids) {
+        std::vector<range_t> instr_ranges;
+        for (const auto &times : workload.times_of_interest) {
+            uint64_t instr_start = 0, instr_end = 0;
+            bool has_start = time_tree_lookup(time_tree[tid_it.second],
+                                              times.start_timestamp, instr_start);
+            bool has_end;
+            if (times.stop_timestamp == 0)
+                has_end = true;
+            else {
+                has_end = time_tree_lookup(time_tree[tid_it.second], times.stop_timestamp,
+                                           instr_end);
+            }
+            if (has_start && has_end && instr_start == instr_end) {
+                ++instr_end;
+            }
+            // If !has_start we'll include from 0.  The start timestamp will make it be
+            // scheduled last but there will be no delay if no other thread is available.
+            // If !has_end, instr_end will still be 0 which means the end of the trace.
+            if (instr_start > 0 || instr_end > 0) {
+                if (!instr_ranges.empty() &&
+                    (instr_ranges.back().stop_instruction >= instr_start ||
+                     instr_ranges.back().stop_instruction == 0)) {
+                    error_string_ =
+                        "times_of_interest are too close together: "
+                        "corresponding instruction ordinals are overlapping or adjacent";
+                    return STATUS_ERROR_INVALID_PARAMETER;
+                }
+                instr_ranges.emplace_back(instr_start, instr_end);
+                VPRINT(this, 3,
+                       "tid %" PRIu64 " overlaps with times_of_interest [%" PRIu64
+                       ", %" PRIu64 ") @ [%" PRIu64 ", %" PRIu64 ")\n",
+                       tid_it.first, times.start_timestamp, times.stop_timestamp,
+                       instr_start, instr_end);
+            }
+        }
+        if (instr_ranges.empty()) {
+            // Exclude this thread completely.  We've already created its
+            // inputs_ entry with cross-indices stored in other structures
+            // so instead of trying to erase it we mark it invalid.
```

Comment:
This ends up complicating interfaces where the analyzer asks about the shard count or asks about the nth input.  Yet erasing has its own issues noted here.  (We assume the analyzer/user can't pass in N inputs plus times_of_interest and expect N shards/final inputs...right?). Suggestions?

Reply:
Don't we have the same issue if the user specifies instr count based ROIs that end up excluding some threads? Do we do anything to handle that scenario?

Reply:
You mean like a really high -skip_instrs?  I think that marks inputs as eof if they don't go that high.  I wonder if it hits the same internal issue I'm hitting where there are no per-shard results because a thread was never seen, causing a crash when generating per-shard results.  But if I remove the whole "valid" and detection of no overlap and try to go the same route: I don't know the instr count so I'd have to set a numeric-limit skip or sthg and have the input go through the initial scheduling on some output but then immediately hitting eof. 
 Feels inefficient.  But if that ends up working maybe that is the simplest.

Reply:
If that also hits that per-shard-gen crash: I'll file that separately as it's orthogonal.

Reply:
Confirmed both things: A) Ripping out this "valid" concept and just specifying a max-limit start works well: that input is scheduled but it hits eof before the user sees it; and B) A high -skip_instrs would hit the same thing: but there's a usage error that prevents it.  A solution that checks never-seen inputs is pretty simple and I'll put that in the 3rd-party code.

Reply:
Done (removed the input_info_t.valid field in favor of an ROI with a max-limit start).